substrate/frame/metadata-hash-extension/src/lib.rs

```diff
@@ -0,0 +1,181 @@
+pub struct EncodeNoneToEmpty(pub Option<[u8; 32]>);
+
+impl Encode for EncodeNoneToEmpty {
+	fn encode(&self) -> Vec<u8> {
+		match self.0 {
+			Some(hash) => hash.encode(),
+			None => Vec::new(),
+		}
+	}
+}
+
+impl TypeInfo for EncodeNoneToEmpty {
+	type Identity = <[u8; 32] as TypeInfo>::Identity;
+
+	fn type_info() -> scale_info::Type {
+		<[u8; 32]>::type_info()
+	}
+}
+
+/// The mode of [`CheckMetadataHash`].
+#[derive(Decode, Encode, PartialEq, Debug, TypeInfo, Clone, Copy, Eq)]
+enum Mode {
+	Disabled,
+	Enabled,
+}
+
+/// Wrapper around the metadata hash and from where to get it from.
+#[derive(Default, Debug, PartialEq, Clone, Copy, Eq)]
+enum MetadataHash {
+	/// Fetch it from the `RUNTIME_METADATA_HASH` env variable at compile time.
+	#[default]
+	FetchFromEnv,
+	/// Use the given metadata hash.
+	Custom([u8; 32]),
+}
+
+impl MetadataHash {
+	/// Returns the metadata hash.
+	fn hash(&self) -> Option<[u8; 32]> {
+		match self {
+			Self::FetchFromEnv =>
+				option_env!("RUNTIME_METADATA_HASH").map(array_bytes::hex2array_unchecked),
+			Self::Custom(hash) => Some(*hash),
+		}
+	}
+}
+
+/// Extension for optionally verifying the metadata hash.
+///
+/// The metadata hash is cryptographical representation of the runtime metadata. This metadata hash
+/// is build as described in [RFC78](https://polkadot-fellows.github.io/RFCs/approved/0078-merkleized-metadata.html).
+/// This metadata hash should give users the confidence that what they build with an online wallet
+/// is the same they are signing with their offline wallet and then applying on chain. To ensure
+/// that the online wallet is not tricking the offline wallet into decoding and showing an incorrect
+/// extrinsic, the offline wallet will include the metadata hash into the additional signed data and
+/// the runtime will then do the same. If the metadata hash doesn't match, the signature
+/// verification will fail and thus, the transaction will be rejected. The RFC contains more details
+/// on how it works.
+///
+/// The extension adds one byte (the `mode`) to the size of the extrinsic. This one byte is
+/// controlling if the metadata hash should be added to the signed data or not. Mode `0` means that
+/// the metadata hash is not added and `1` means that it is added. Further values of `mode` are
+/// reserved for future changes.
+///
+/// The metadata hash is read from the environment variable `RUNTIME_METADATA_HASH`. This
+/// environment variable is for example set by the `substrate-wasm-builder` when the feature for
+/// generating the metadata hash is enabled. If the environment variable is not set and `mode = 1`
+/// is passed, the transaction is rejected with [`UnknownTransaction::CannotLookup`].
+#[derive(Encode, Decode, Clone, Eq, PartialEq, TypeInfo, DebugNoBound)]
+#[scale_info(skip_type_params(T))]
+pub struct CheckMetadataHash<T> {
+	_phantom: core::marker::PhantomData<T>,
+	mode: Mode,
+	#[codec(skip)]
+	metadata_hash: MetadataHash,
```

Comment:
May be worth having a type parameter to choose between `FetchFromEnv` and `Custom`. Now you can go with custom only in tests.

Reply:
If you use `CheckMetadataHash::new` you will get `FetchFromEnv`.